cloudinit/sources/DataSourceEc2.py

```diff
@@ -947,6 +948,85 @@ def _get_key_as_int_or(dikt, key, alt_value):
     }
 
 
+def _configure_policy_routing(
+    dev_config: dict,
+    *,
+    nic_name: str,
+    nic_metadata: dict,
+    distro: Distro,
+    is_ipv4: bool,
+    table: int,
+) -> None:
+    """
+    Configure policy-based routing on secondary NICs / secondary IPs to
+    ensure outgoing packets are routed via the correct interface.
+
+    @param: dev_config: network cfg v2 to be updated inplace.
+    @param: nic_name: nic name. Only used if ipv4.
+    @param: nic_metadata: nic metadata from IMDS.
+    @param: distro: Instance of Distro. Only used if ipv4.
+    @param: is_ipv4: Boolean indicating if we are acting over ipv4 or not.
+    @param: table: Routing table id.
+    """
+    if not dev_config.get("routes"):
+        dev_config["routes"] = []
+    if is_ipv4:
+        try:
+            lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
+            gateway = lease["routers"]
+        except NoDHCPLeaseError as e:
+            LOG.warning(
+                "Could not perform dhcp discovery on %s to find its "
+                "gateway. Not adding default route via the gateway. "
+                "Error: %s",
+                nic_name,
+                e,
+            )
+        else:
+            # Add default route via the NIC's gateway
+            dev_config["routes"].append(
+                {
+                    "to": "0.0.0.0/0",
+                    "via": gateway,
+                    "table": table,
+                },
+            )
+
+    if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
+    else:
+        subnet_prefix_routes = nic_metadata["subnet-ipv6-cidr-blocks"]
+    subnet_prefix_routes = (
+        [subnet_prefix_routes]
+        if isinstance(subnet_prefix_routes, str)
+        else subnet_prefix_routes
+    )
+    for prefix_route in subnet_prefix_routes:
+        dev_config["routes"].append(
+            {
+                "to": prefix_route,
+                "table": table,
+            },
+        )
+
+    if not dev_config.get("routing-policy"):
+        dev_config["routing-policy"] = []
+    # Packets coming from any IP associated with the current NIC
+    # will be routed using `table` routing table
+    if is_ipv4:
+        ips = nic_metadata["local-ipv4s"]
+    else:
+        ips = nic_metadata["ipv6s"]
```

Comment:
Can't this conditional also be pulled into that first `is is_ipv4 too?
```diff
--- a/cloudinit/sources/DataSourceEc2.py
+++ b/cloudinit/sources/DataSourceEc2.py
@@ -971,6 +971,8 @@ def _configure_policy_routing(
     if not dev_config.get("routes"):
         dev_config["routes"] = []
     if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
+        ips = nic_metadata["local-ipv4s"]
         try:
             lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
             gateway = lease["routers"]
@@ -991,11 +993,10 @@ def _configure_policy_routing(
                     "table": table,
                 },
             )
-
-    if is_ipv4:
-        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
     else:
         subnet_prefix_routes = nic_metadata["subnet-ipv6-cidr-blocks"]
+        ips = nic_metadata["ipv6s"]
+
     subnet_prefix_routes = (
         [subnet_prefix_routes]
         if isinstance(subnet_prefix_routes, str)
@@ -1013,10 +1014,6 @@ def _configure_policy_routing(
         dev_config["routing-policy"] = []
     # Packets coming from any IP associated with the current NIC
     # will be routed using `table` routing table
-    if is_ipv4:
-        ips = nic_metadata["local-ipv4s"]
-    else:
-        ips = nic_metadata["ipv6s"]
     ips = [ips] if isinstance(ips, str) else ips
     for ip in ips:
         dev_config["routing-policy"].append(
```

Reply:
Applied, plus additional reference to GH issue.